### ISSUE TITLE
A problem with cross product: if second column has more than 4 entries results are incorrect

### ISSUE THREAD

**Issue by iguberman** [joergen7/cuneiform#38]
I noticed it during testing 24 hours by 6 chunks.   
Then I created a little test to see what's going on:
```
deftask gen-xx-sequence( <items(String)> : last ) in bash *{
items=`for item in \`seq 0 $last\`; do printf '%02d\n' $item; done`
}*
deftask process-hour-minute( out(File) : hour minute ) in bash *{
echo "HH:$hour MM:$minute" > $out
}*
deftask join-files( out : <files(File)> ) in bash *{
cat ${files[@]) > $out
}*
files = process-hour-minute( hour : gen-xx-sequence(last : 23), minute : gen-xx-sequence( last : 59));
result = join-files( files : files );
result;
```

Seems like number 4 has some magic powers (in regards to second column) when crossing two columns.  I looked at the code and there is only one place where 4 is hard coded, in LocalThread:
```
private static final int MAX_TRIALS = 4;
...
 } while( suc == false && trial <= MAX_TRIALS );
```

Here are a few examples:

I.e. if I sent only 10 hours (0-9) and 5 minutes (0-4), I get  (9 * 4) + (1 * 5) (because the first 9 hours only get 4 minutes and the last one gets all 5)

HH:00 MM:00  <-- hours 00-08 only get paired with 4 minutes instead of 5
HH:00 MM:01
HH:00 MM:02
HH:00 MM:03
HH:01 MM:00
HH:01 MM:01
HH:01 MM:02
HH:01 MM:03
HH:02 MM:00
HH:02 MM:01
HH:02 MM:02
HH:02 MM:03
HH:03 MM:00
HH:03 MM:01
HH:03 MM:02
HH:03 MM:03
HH:04 MM:00
HH:04 MM:01
HH:04 MM:02
HH:04 MM:03
HH:05 MM:00
HH:05 MM:01
HH:05 MM:02
HH:05 MM:03
HH:06 MM:00
HH:06 MM:01
HH:06 MM:02
HH:06 MM:03
HH:07 MM:00
HH:07 MM:01
HH:07 MM:02
HH:07 MM:03
HH:08 MM:00
HH:08 MM:01
HH:08 MM:02
HH:08 MM:03
**HH:09 MM:00   <-- Last one is correctly paired with 5 "minute" entries
HH:09 MM:01
HH:09 MM:02
HH:09 MM:03
HH:09 MM:04**

If I send 5 hours (0-4) and 10 minutes (0-9) I get: (4 * 4) + 1 * 10  instead of 5 * 10
`>for i in `ls ~/.cuneiform/repo`; do cat ~/.cuneiform/repo/$i; done`

HH:00 MM:00     <- 0th hour only gets 4 minute pairing instead of 10
HH:00 MM:01
HH:00 MM:02
HH:00 MM:03
HH:01 MM:00     <- 1st hour only gets 4 minute pairing instead of 10
HH:01 MM:01
HH:01 MM:02
HH:01 MM:03
HH:02 MM:00    <- 2nd hour only gets 4 minute pairing instead of 10
HH:02 MM:01
HH:02 MM:02
HH:02 MM:03
HH:03 MM:00    <- 3rd hour only gets 4 minute pairing instead of 10
HH:03 MM:01
HH:03 MM:02
HH:03 MM:03
HH:04 MM:00    <- Last one is correctly paired with 10 "minute" entries
HH:04 MM:01
HH:04 MM:02
HH:04 MM:03
HH:04 MM:04
HH:04 MM:05
HH:04 MM:06
HH:04 MM:07
HH:04 MM:08
HH:04 MM:09

Results get more interesting with a larger number on the "minute" column.  Here I use 0-2 hours and 0-49 minutes, expecting 150 results, but got only 90,  (again, the last entry **hour 02** looks good, but for the first two "magic number" 4 still shows up here, but now it repeats)
for i in `ls ~/.cuneiform/repo`; do cat ~/.cuneiform/repo/$i; done | grep -c HH
90
➜  CUNEIFORM_EXPERIMENT1  for i in `ls ~/.cuneiform/repo`; do cat ~/.cuneiform/repo/$i; done             
HH:00 MM:00
HH:00 MM:01
HH:00 MM:02
HH:00 MM:03
HH:01 MM:00
HH:01 MM:01
HH:01 MM:02
HH:01 MM:03
**HH:02 MM:00
HH:02 MM:01
HH:02 MM:02
HH:02 MM:03
HH:02 MM:04
HH:02 MM:05
HH:02 MM:06
HH:02 MM:07
HH:02 MM:08
HH:02 MM:09**
HH:00 MM:10
HH:00 MM:11
HH:00 MM:12
HH:00 MM:13
HH:01 MM:10
HH:01 MM:11
HH:01 MM:12
HH:01 MM:13
**HH:02 MM:10
HH:02 MM:11
HH:02 MM:12
HH:02 MM:13
HH:02 MM:14
HH:02 MM:15
HH:02 MM:16
HH:02 MM:17
HH:02 MM:18
HH:02 MM:19**
HH:00 MM:20
HH:00 MM:21
HH:00 MM:22
HH:00 MM:23
HH:01 MM:20
HH:01 MM:21
HH:01 MM:22
HH:01 MM:23
**HH:02 MM:20
HH:02 MM:21
HH:02 MM:22
HH:02 MM:23
HH:02 MM:24
HH:02 MM:25
HH:02 MM:26
HH:02 MM:27
HH:02 MM:28
HH:02 MM:29**
HH:00 MM:30
HH:00 MM:31
HH:00 MM:32
HH:00 MM:33
HH:01 MM:30
HH:01 MM:31
HH:01 MM:32
HH:01 MM:33
**HH:02 MM:30
HH:02 MM:31
HH:02 MM:32
HH:02 MM:33
HH:02 MM:34
HH:02 MM:35
HH:02 MM:36
HH:02 MM:37
HH:02 MM:38
HH:02 MM:39**
HH:00 MM:40
HH:00 MM:41
HH:00 MM:42
HH:00 MM:43
HH:01 MM:40
HH:01 MM:41
HH:01 MM:42
HH:01 MM:43
**HH:02 MM:40
HH:02 MM:41
HH:02 MM:42
HH:02 MM:43
HH:02 MM:44
HH:02 MM:45
HH:02 MM:46
HH:02 MM:47
HH:02 MM:48
HH:02 MM:49**

Note, if I send 10 hours and only 4 minutes (to fit in with the magic number):
files = process-hour-minute( hour : gen-xx-sequence(last : 9 ), minute : gen-xx-sequence( last : 3));
Yes, now I got 10 * 4:  
 for i in `ls ~/.cuneiform/repo`; do cat ~/.cuneiform/repo/$i; done
HH:00 MM:00
HH:00 MM:01
HH:00 MM:02
HH:00 MM:03
HH:01 MM:00
HH:01 MM:01
HH:01 MM:02
HH:01 MM:03
HH:02 MM:00
HH:02 MM:01
HH:02 MM:02
HH:02 MM:03
HH:03 MM:00
HH:03 MM:01
HH:03 MM:02
HH:03 MM:03
HH:04 MM:00
HH:04 MM:01
HH:04 MM:02
HH:04 MM:03
HH:05 MM:00
HH:05 MM:01
HH:05 MM:02
HH:05 MM:03
HH:06 MM:00
HH:06 MM:01
HH:06 MM:02
HH:06 MM:03
HH:07 MM:00
HH:07 MM:01
HH:07 MM:02
HH:07 MM:03
HH:08 MM:00
HH:08 MM:01
HH:08 MM:02
HH:08 MM:03
HH:09 MM:00
HH:09 MM:01
HH:09 MM:02
HH:09 MM:03

So if I have a list of no more than 4 in second column --> I get correct number of cross entries.

**Comment by iguberman** [comment]
Actually, I just tried a set with 24 hours and 3 "chunks": 1 2 3 and ended up with 20 of chunk 1, 20 of chunk 2 and 24 of chunk 3 (64 total instead of 72).   Having less than 4 items in second column isn't a workaround.   Unless I'm doing something terribly wrong on my end, this seems like a rather huge problem and should be high priority to fix.  This problem is consistent with either local or htcondor backend.   Also, the cross-product "math" seems very consistent: I always get 24X3=64 no matter how many times I run.     Looks like it is very basic behavior that's being compromised.   I'm following the cuneiform's example from http://ceur-ws.org/Vol-1330/paper-03.pdf (temp ph cross product example)    Also, I'm currently using 2.0.3-RELEASE version of CF.

**Comment by iguberman** [comment]
P.S. Also, in my test above I generate 24 hours in a script.   That's not a problem.   I ran with hours = 0 1 2 3...  before I got the sequence-generation fixed on my end.

**Comment by joergen7** [comment]
It is very good that you point that out. This is really a bad issue. 
I'll have a look at it as soon as possible.

You'll hear from me on monday about it, whether I solved the problem or not.

Cheers
Jörgen

**Comment by joergen7** [comment]
I looked at this a bit closer now. The enumeration of the cross product seems to be ok, it is the hash value that is generated for each task invocation that makes problems.

The cache uses these hashes to determine whether it has already computed a result or not. I thought, collisions in this hash value were extremely seldom but apparently they aren't.

I'm working on it...

**Comment by joergen7** [comment]
I have created a fix which you can find in the current `dev` branch of the cuneiform repository. Please be sure to execute the 2016-02-08 build and keep me updated.

**Comment by iguberman** [comment]
Awesome!   It is a good problem to get rid of! :)


The funny thing is I may not even need two columns in my project.    We had
problems with files being transferred to the worker for joining because the
DAG in cuneiform is one for all.   So I made some functional changes in my
version of 2.0.3-RELEASE (because I needed it in a stable version).

So, I'll be creating a pull requests for some of the modifications I made
in the dev branch for 2.0.4-SNAPSHOT.  Maybe you'll like them to be in
2.0.4 if you agree they are important (as a heavy condor user I think they
are).

Also, I was able to fix htcondor workflow not shutting down as well.

We'll try to get them to you for Code Review hopefully later tonight.

Cheers,
Irina.
